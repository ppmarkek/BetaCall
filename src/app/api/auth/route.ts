import axios from 'axios';

const authURL = 'https://localhost:8080/api/users';

export const userSignIn = async ({
  email,
  password,
}: {
  email: string;
  password: string;
}) => {
  try {
    const response = await axios.post(`${authURL}/login`, { email, password });
    return response.data;
  } catch (err) {
    return console.error(err);
  }
};
