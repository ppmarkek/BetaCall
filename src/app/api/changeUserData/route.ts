import axios, { AxiosResponse } from 'axios';

const URL =
  'https://betacall-backend.onrender.com/api/users/request-reset-password';

export const requestResetPassword = async (email: string) => {
  try {
    const response = await axios.post(URL, { email });
    return response;
  } catch (err) {
    if (axios.isAxiosError(err) && err.response) {
      return err.response as AxiosResponse<unknown>;
    }
    throw err;
  }
};