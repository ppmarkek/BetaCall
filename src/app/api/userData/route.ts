import axios, { AxiosResponse } from 'axios';

const resetPasswordURL =
  'https://betacall-backend.onrender.com/api/users/request-reset-password';

export const tokenResetPassword = async (
  newPassword: string,
  token: string
) => {
  try {
    const response = await axios.post(`${resetPasswordURL}/${token}`, {
      newPassword,
    });
    return response;
  } catch (err) {
    if (axios.isAxiosError(err) && err.response) {
      return err.response as AxiosResponse<unknown>;
    }
    throw err;
  }
};

export const requestResetPassword = async (email: string) => {
  try {
    const response = await axios.post(resetPasswordURL, { email });
    return response;
  } catch (err) {
    if (axios.isAxiosError(err) && err.response) {
      return err.response as AxiosResponse<unknown>;
    }
    throw err;
  }
};
