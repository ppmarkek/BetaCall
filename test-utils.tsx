import { ReactElement, FC } from 'react';
import { ChakraProvider, defaultSystem } from '@chakra-ui/react';
import { render, RenderOptions } from '@testing-library/react';

const Wrapper: FC<{ children: React.ReactNode }> = ({ children }) => (
  <ChakraProvider value={defaultSystem}>{children}</ChakraProvider>
);

const customRender = (ui: ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: Wrapper, ...options });

export * from '@testing-library/react';
export { customRender as render };
