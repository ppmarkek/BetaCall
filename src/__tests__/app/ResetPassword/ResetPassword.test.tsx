import React from 'react';
import { fireEvent, render, screen, waitFor } from '../../../../test-utils';
import '@testing-library/jest-dom';
import ResetPasswordPage from '@/app/resetPassword/[token]/page';
import { tokenResetPassword } from '@/app/api/userData/route';
import { AxiosResponse, AxiosHeaders } from 'axios';
import * as nextNavigation from 'next/navigation';

jest.mock('@/app/api/userData/route');

jest.mock('next/navigation', () => ({
  useParams: jest.fn(() => ({ token: 'test-token' })),
}));

const emptyHeaders = new AxiosHeaders();

describe('Reset Password Page', () => {
  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('renders the reset password page', () => {
    render(<ResetPasswordPage />);
    expect(screen.getByTestId('ResetPasswordPage-testid')).toBeInTheDocument();
  });

  it('check new password visibility', () => {
    render(<ResetPasswordPage />);
    const newPasswordInput = screen.getByPlaceholderText(
      'Enter your new password'
    ) as HTMLInputElement;

    expect(newPasswordInput.type).toBe('password');

    const icons = screen.getAllByTestId('icon-textInput');
    const newPasswordToggleIcon = icons[0];
    fireEvent.click(newPasswordToggleIcon);

    expect(newPasswordInput.type).toBe('text');
  });

  it('check confirm new password visibility', () => {
    render(<ResetPasswordPage />);
    const confirmPasswordInput = screen.getByPlaceholderText(
      'Confirm your new password'
    ) as HTMLInputElement;

    expect(confirmPasswordInput.type).toBe('password');

    const icons = screen.getAllByTestId('icon-textInput');
    const confirmPasswordToggleIcon = icons[1];
    fireEvent.click(confirmPasswordToggleIcon);

    expect(confirmPasswordInput.type).toBe('text');
  });

  it('shows spinner on form submit and hides it after request resolves', async () => {
    let resolveRequest!: (value: AxiosResponse<{ message: string }>) => void;
    const requestPromise = new Promise<AxiosResponse<{ message: string }>>(
      (resolve) => {
        resolveRequest = resolve;
      }
    );

    (tokenResetPassword as jest.Mock).mockReturnValue(requestPromise);

    render(<ResetPasswordPage />);

    const newPasswordInput = screen.getByPlaceholderText(
      'Enter your new password'
    );
    fireEvent.change(newPasswordInput, {
      target: { value: '{S[6O$}V56v^' },
    });

    const newConfirmPasswordInput = screen.getByPlaceholderText(
      'Confirm your new password'
    );
    fireEvent.change(newConfirmPasswordInput, {
      target: { value: '{S[6O$}V56v^' },
    });

    const submitButton = screen.getByTestId('button-submit');
    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(screen.getByTestId('spinner')).toBeInTheDocument();
    });

    resolveRequest({
      data: { message: 'Password updated successfully' },
      status: 200,
      statusText: 'OK',
      headers: emptyHeaders,
      config: { headers: emptyHeaders },
    });

    await waitFor(() => {
      expect(screen.queryByTestId('spinner')).not.toBeInTheDocument();
    });

    expect(
      screen.getByText('Password updated successfully')
    ).toBeInTheDocument();
  });

  it('set message when bad request', async () => {
    let resolveRequest!: (value: AxiosResponse<{ message: string }>) => void;
    const requestPromise = new Promise<AxiosResponse<{ message: string }>>(
      (resolve) => {
        resolveRequest = resolve;
      }
    );

    (tokenResetPassword as jest.Mock).mockReturnValue(requestPromise);

    render(<ResetPasswordPage />);

    const newPasswordInput = screen.getByPlaceholderText(
      'Enter your new password'
    );
    fireEvent.change(newPasswordInput, {
      target: { value: '{S[6O$}V56v^' },
    });

    const newConfirmPasswordInput = screen.getByPlaceholderText(
      'Confirm your new password'
    );
    fireEvent.change(newConfirmPasswordInput, {
      target: { value: '{S[6O$}V56v^' },
    });

    const submitButton = screen.getByTestId('button-submit');
    fireEvent.click(submitButton);

    resolveRequest({
      data: { message: 'Invalid or expired password reset token' },
      status: 400,
      statusText: 'Not Found',
      headers: emptyHeaders,
      config: { headers: emptyHeaders },
    });

    await waitFor(() => {
      expect(
        screen.getByText('Invalid or expired password reset token')
      ).toBeInTheDocument();
    });
  });

  it('calls console.error with "Token is missing" error when token is missing', async () => {
    const useParamsMock = jest
      .spyOn(nextNavigation, 'useParams')
      .mockReturnValue({ token: undefined } as { token?: string });

    const consoleErrorSpy = jest
      .spyOn(console, 'error')
      .mockImplementation(() => {});

    render(<ResetPasswordPage />);

    const validPassword = '{S[6O$}V56v^';
    const newPasswordInput = screen.getByPlaceholderText(
      'Enter your new password'
    );
    fireEvent.change(newPasswordInput, { target: { value: validPassword } });

    const confirmPasswordInput = screen.getByPlaceholderText(
      'Confirm your new password'
    );
    fireEvent.change(confirmPasswordInput, {
      target: { value: validPassword },
    });

    const submitButton = screen.getByTestId('button-submit');
    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(consoleErrorSpy).toHaveBeenCalled();
    });

    const errorLogged = consoleErrorSpy.mock.calls[0][0];
    expect(errorLogged).toBeInstanceOf(Error);
    expect(errorLogged.message).toEqual('Token is missing');

    useParamsMock.mockRestore();
    consoleErrorSpy.mockRestore();
  });

  it('matches the snapshot', () => {
    (nextNavigation.useParams as jest.Mock).mockReturnValue({
      token: 'test-token',
    });
    const { container } = render(<ResetPasswordPage />);
    expect(container).toMatchSnapshot();
  });
});
