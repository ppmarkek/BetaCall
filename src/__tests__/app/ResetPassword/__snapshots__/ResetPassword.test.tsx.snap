// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reset Password Page matches the snapshot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: calc(100svh - 81px);
  min-height: 800px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 420px;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  text-align: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 50px;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 10px;
}

.emotion-3 {
  font-family: Lato;
  font-weight: 700;
  font-size: 32px;
  line-height: 42px;
}

.emotion-4 {
  color: #8083A3;
  font-family: Lato;
  font-size: 14px;
  font-weight: 400;
  line-height: 21px;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 35px;
  width: 420px;
}

.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 10px;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  position: relative;
}

.emotion-6 .input-icon {
  fill: #8083a3;
  -webkit-transition: fill 0.3s;
  transition: fill 0.3s;
}

.emotion-6:focus-within .input-icon:not(.input-icon-error) {
  fill: #1a1c1d;
}

.emotion-6 .input-icon-error {
  fill: #ff808b;
}

.emotion-7 {
  color: #8083A3;
  font-family: Lato;
  font-size: 16px;
  font-weight: 400;
  line-height: 19px;
}

.emotion-8 {
  color: #1a1c1d;
  padding: 20px 0px;
  padding-left: 0px;
  padding-right: 25px;
  border: unset;
  border-radius: unset;
  border-bottom: 1px solid #eceef5;
  color: #1a1c1d;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
}

@layer recipes {
  .emotion-8 {
    width: 100%;
    min-width: var(--input-height);
    outline: 0;
    position: relative;
    -webkit-appearance: none;
    -moz-appearance: none;
    -ms-appearance: none;
    appearance: none;
    text-align: start;
    border-radius: var(--chakra-radii-l2);
    height: var(--input-height);
    --focus-color: var(--chakra-colors-color-palette-focus-ring);
    --error-color: var(--chakra-colors-border-error);
    font-size: var(--chakra-font-sizes-sm);
    line-height: 1.25rem;
    padding-inline: var(--chakra-spacing-3);
    --input-height: var(--chakra-sizes-10);
    background: var(--chakra-colors-transparent);
    --bg-currentcolor: var(--chakra-colors-transparent);
    border-width: 1px;
    border-color: var(--chakra-colors-border);
    --focus-ring-color: var(--chakra-colors-color-palette-focus-ring);
  }

  .emotion-8:is(:disabled, [disabled], [data-disabled], [aria-disabled=true]) {
    opacity: 0.5;
    cursor: not-allowed;
  }

  .emotion-8:is([data-invalid], [aria-invalid=true], [data-state=invalid]) {
    --focus-ring-color: var(--error-color);
    border-color: var(--error-color);
  }

  .emotion-8:is(:focus-visible, [data-focus-visible]) {
    outline-offset: 0px;
    outline-width: var(--focus-ring-width, 1px);
    outline-color: var(--focus-ring-color);
    outline-style: var(--focus-ring-style, solid);
    border-color: var(--focus-ring-color);
  }
}

.emotion-8:focus-visible {
  outline: unset;
  border-bottom: 1px solid #000;
}

.emotion-9 {
  position: absolute;
  right: 0;
  bottom: 8px;
  font-size: 20px;
  cursor: pointer;
}

@layer recipes {
  .emotion-9 {
    display: inline-block;
    line-height: 1em;
    -webkit-flex-shrink: 0;
    -ms-flex-negative: 0;
    flex-shrink: 0;
    color: currentcolor;
    vertical-align: middle;
  }
}

.emotion-14 {
  background-color: #6B59CC;
  --bg-currentcolor: #6B59CC;
  height: 50px;
  font-size: 16px;
  font-weight: 700;
  line-height: 20px;
  font-family: 'Lato',sans-serif;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 10px;
  padding: 0 30px;
  border-radius: 10px;
  width: 100%;
  border: 1px solid #6B59CC;
  background: #6B59CC;
  color: #fff;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
  cursor: pointer;
}

@layer recipes {
  .emotion-14 {
    display: -webkit-inline-box;
    display: -webkit-inline-flex;
    display: -ms-inline-flexbox;
    display: inline-flex;
    -webkit-appearance: none;
    -moz-appearance: none;
    -ms-appearance: none;
    appearance: none;
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    -webkit-box-pack: center;
    -ms-flex-pack: center;
    -webkit-justify-content: center;
    justify-content: center;
    -webkit-user-select: none;
    -moz-user-select: none;
    -ms-user-select: none;
    user-select: none;
    position: relative;
    border-radius: var(--chakra-radii-l2);
    white-space: nowrap;
    vertical-align: middle;
    border-width: 1px;
    border-color: var(--chakra-colors-transparent);
    cursor: var(--chakra-cursor-button);
    -webkit-flex-shrink: 0;
    -ms-flex-negative: 0;
    flex-shrink: 0;
    outline: 0;
    line-height: 1.25rem;
    isolation: isolate;
    font-weight: var(--chakra-font-weights-medium);
    transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,translate,transform;
    transition-duration: var(--chakra-durations-moderate);
    --focus-ring-color: var(--chakra-colors-color-palette-focus-ring);
    height: var(--chakra-sizes-10);
    min-width: var(--chakra-sizes-10);
    font-size: var(--chakra-font-sizes-sm);
    padding-inline: var(--chakra-spacing-4);
    gap: var(--chakra-spacing-2);
    background: var(--chakra-colors-color-palette-solid);
    --bg-currentcolor: var(--chakra-colors-color-palette-solid);
    color: var(--chakra-colors-color-palette-contrast);
  }

  .emotion-14:is(:focus-visible, [data-focus-visible]) {
    outline-width: var(--focus-ring-width, 2px);
    outline-offset: var(--focus-ring-offset, 2px);
    outline-style: var(--focus-ring-style, solid);
    outline-color: var(--focus-ring-color);
  }

  .emotion-14:is(:disabled, [disabled], [data-disabled], [aria-disabled=true]) {
    opacity: 0.5;
    cursor: not-allowed;
  }

  .emotion-14 :where(svg) {
    -webkit-flex-shrink: 0;
    -ms-flex-negative: 0;
    flex-shrink: 0;
    width: var(--chakra-sizes-5);
    height: var(--chakra-sizes-5);
  }

  .emotion-14:is([aria-expanded=true], [data-expanded], [data-state=expanded]) {
    --mix-background: color-mix(in srgb, var(--chakra-colors-color-palette-solid) 90%, transparent);
    background: var(--mix-background, var(--chakra-colors-color-palette-solid));
    --bg-currentcolor: var(--mix-background, var(--chakra-colors-color-palette-solid));
  }

  @media (hover: hover) {
    .emotion-14:is(:hover, [data-hover]):not(:disabled, [data-disabled]) {
      --mix-background: color-mix(in srgb, var(--chakra-colors-color-palette-solid) 90%, transparent);
      background: var(--mix-background, var(--chakra-colors-color-palette-solid));
      --bg-currentcolor: var(--mix-background, var(--chakra-colors-color-palette-solid));
    }
  }
}

.emotion-14 svg {
  fill: #fff;
  -webkit-transition: fill 0.3s;
  transition: fill 0.3s;
}

.emotion-14:focus-visible {
  box-shadow: none!important;
  outline: none!important;
}

.emotion-14:hover {
  background: unset;
  color: #6B59CC;
}

.emotion-14:hover svg {
  fill: #6B59CC;
}

<div>
  <div
    class="emotion-0"
    data-testid="ResetPasswordPage-testid"
  >
    <div
      class="emotion-1"
    >
      <div
        class="emotion-2"
      >
        <p
          class="emotion-3"
        >
          Reset Your Password
        </p>
        <p
          class="emotion-4"
        >
          Enter your new password below
        </p>
      </div>
      <form>
        <div
          class="emotion-5"
        >
          <div
            class="emotion-6"
          >
            <p
              class="emotion-7"
            >
              New Password
            </p>
            <input
              class="chakra-input chakra-tooltip__trigger emotion-8"
              data-part="trigger"
              data-scope="tooltip"
              data-state="closed"
              dir="ltr"
              id="tooltip::rc::trigger"
              name="newPassword"
              placeholder="Enter your new password"
              type="password"
            />
            <svg
              aria-hidden="true"
              class="chakra-icon input-icon emotion-9"
              data-testid="icon-textInput"
              fill="currentColor"
              focusable="false"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zm-6 9c-1.1 0-2-.9-2-2s.9-2 2-2 2 .9 2 2-.9 2-2 2zm3.1-9H8.9V6c0-1.71 1.39-3.1 3.1-3.1 1.71 0 3.1 1.39 3.1 3.1v2z"
              />
            </svg>
          </div>
          <div
            class="emotion-6"
          >
            <p
              class="emotion-7"
            >
              Confirm New Password
            </p>
            <input
              class="chakra-input chakra-tooltip__trigger emotion-8"
              data-part="trigger"
              data-scope="tooltip"
              data-state="closed"
              dir="ltr"
              id="tooltip::rd::trigger"
              name="repeatNewPassword"
              placeholder="Confirm your new password"
              type="password"
            />
            <svg
              aria-hidden="true"
              class="chakra-icon input-icon emotion-9"
              data-testid="icon-textInput"
              fill="currentColor"
              focusable="false"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
              />
            </svg>
          </div>
          <button
            class="chakra-button emotion-14"
            data-testid="button-submit"
            type="submit"
          >
            Save
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
